src/Storages/AlterCommands.cpp

```diff
@@ -1151,8 +1156,10 @@ void AlterCommands::validate(const StoragePtr & table, ContextPtr context) const
                 const GetColumnsOptions options(GetColumnsOptions::AllPhysical);
                 const auto old_data_type = all_columns.getColumn(options, column_name).type;
 
-                if (command.data_type->getName().contains("Object")
-                    || old_data_type->getName().contains("Object"))
+                bool new_type_is_object = !!typeid_cast<const DataTypeObject *>(command.data_type.get());
```

Comment:
You can do something stupid like:
```
ALTER TABLE testJson
    ADD COLUMN `j2` Tuple(JSON) DEFAULT jString
```

Not sure how that's meant to work, but even with these changes it's still allowed.

Reply:
Okay, let me traverse the tree of types.

Reply:
Oh, we already have `hasDynamicSubcolumns`.